std/range/package.d

```diff
@@ -5887,284 +5887,387 @@ pure @safe nothrow unittest
+            end = cast(E) (end - 1);
+            end -= unsigned(end - begin) % step;
         }
-
-        @property inout(Value) back() inout
+        else if (begin > end && step < 0)
         {
-            assert(!empty);
-            return last;
+            // Iterating downward
+            assert(unsigned((begin - end) / (0 - step)) <= size_t.max);
+            // Cast below can't fail because begin > end
+            end = cast(E) (end + 1);
+            end += unsigned(begin - end) % (0 - step);
         }
-        void popBack()
+        else
         {
-            assert(!empty);
-            if (current == last) step = 0;
-            else last -= step;
+            // Initialize an empty range
+            step = 0;
+            return;
         }
+    }
 
-        @property auto save() { return this; }
+    @property
+    auto save() inout { return this; }
 
-        inout(Value) opIndex(ulong n) inout
-        {
-            assert(n < this.length);
+    inout(IotaResult) opSlice() inout { return this; }
 
-            // Just cast to Value here because doing so gives overflow behavior
-            // consistent with calling popFront() n times.
-            return cast(inout Value) (current + step * n);
-        }
-        auto opBinaryRight(string op)(Value val) const
-        if (op == "in")
+    void popFront()
+    {
+        assert(!empty, "Can't popFront on an empty range.");
+        if (begin == end)
+            step = 0;
+        else
+            begin += step;
+    }
+
+    void popBack()()
+    {
+        assert(!empty, "Can't popBack on an empty range.");
+        if (begin == end)
+            step = 0;
+        else
+            end -= step;
+    }
+
+    @property
+    inout(ValueType) front() inout
+    {
+        assert(!empty, "Can't get the front of an empty range.");
+        return begin;
+    }
+
+    @property
+    inout(ValueType) back()() inout
+    {
+        assert(!empty, "Can't get the back of an empty range.");
+        return end;
+    }
+
+    @property
+    bool empty() const
+    {
+        static if (is(typeof(begin < end)))
         {
-            if (empty) return false;
-            //cast to avoid becoming unsigned
-            auto supposedIndex = cast(StepType)(val - current) / step;
-            return supposedIndex < length && supposedIndex * step + current == val;
+            if (step == 0) return true;
+            if (step < 0)
+                return begin < end;
+            return begin > end;
         }
-        auto contains(Value x){return x in this;}
-        inout(Result) opSlice() inout { return this; }
-        inout(Result) opSlice(ulong lower, ulong upper) inout
+        else static if (is(typeof(begin == end)))
         {
-            assert(upper >= lower && upper <= this.length);
+            return begin == end;
+        }
+        else static assert(false, "Arguments to iota must be comparable using > or ==");
+    }
```

Comment:
`empty` seems off, too. A simple invariant that is established by the constructor and popXxx is "a range is empty if and only if the step is zero". Then `empty` would be a one-liner using that invariant.